x/shield/keeper/proposal.go

```diff
@@ -4,16 +4,182 @@ import (
+		return types.ErrNoPoolFound
+	}
+	if lossAmt.GT(pool.Shield) {
+		return types.ErrNotEnoughShield
+	}
+
+	// Verify collateral availability.
+	totalCollateral := k.GetTotalCollateral(ctx)
+	//totalWithdrawing := k.GetTotalWithdrawing(ctx) //TODO: No need for this variable?
+	totalLocked := k.GetTotalLocked(ctx)
+
+	// Ensure that total collateral (withdrawing and non-withdrawing)
+	// can cover the new total lock amount.
+	if totalLocked.Add(lossAmt).GT(totalCollateral) {
+		return types.ErrNotEnoughCollateral
+	}
+
+	// Lock proportional amount from each provider.
+	providers := k.GetAllProviders(ctx)
+	proportion := lossAmt.ToDec().Quo(totalCollateral.ToDec())
+	remaining := lossAmt
+	for i := range providers {
+		var lockAmt sdk.Int
+		if i < len(providers)-1 {
+			lockAmt = providers[i].Collateral.ToDec().Mul(proportion).TruncateInt()
+			if lockAmt.LT(providers[i].Collateral) && lockAmt.LT(remaining) {
+				lockAmt = lockAmt.Add(sdk.OneInt())
+			}
+			remaining = remaining.Sub(lockAmt)
+		} else {
+			lockAmt = remaining
+		}
+		k.LockProvider(ctx, providers[i], types.NewLockedCollateral(proposalID, lockAmt), lockPeriod)
+	}
+
+	// Update shield amount and delete time of the purchase.
+	purchaseList, found := k.GetPurchaseList(ctx, poolID, purchaser)
+	if !found {
+		return types.ErrPurchaseNotFound
+	}
+	var index int
+	for i, entry := range purchaseList.Entries {
+		if entry.PurchaseID == purchaseID {
+			index = i
+			break
+		}
+	}
+	purchase := &purchaseList.Entries[index]
+	if lossAmt.GT(purchase.Shield) {
+		return types.ErrNotEnoughShield
+	}
+	k.DequeuePurchase(ctx, purchaseList, purchase.DeletionTime)
+	purchase.Shield = purchase.Shield.Sub(lossAmt)
+	votingEndTime := ctx.BlockTime().Add(lockPeriod)
+	if purchase.DeletionTime.Before(votingEndTime) {
+		// TODO: correctly update delete time & protection end time
+		purchase.DeletionTime = votingEndTime // temp
+	}
+	k.SetPurchaseList(ctx, purchaseList)
+	k.InsertExpiringPurchaseQueue(ctx, purchaseList, purchase.DeletionTime)
+
+	// Update pool and global pool states
+	pool.Shield = pool.Shield.Sub(lossAmt)
+	k.SetPool(ctx, pool)
+
+	totalLocked = totalLocked.Add(lossAmt)
+	totalCollateral = totalCollateral.Sub(lossAmt)
+	totalShield := k.GetTotalShield(ctx)
+	totalShield = totalShield.Sub(lossAmt)
+	k.SetTotalLocked(ctx, totalLocked)
+	k.SetTotalCollateral(ctx, totalCollateral)
+	k.SetTotalShield(ctx, totalShield)
+
 	return nil
 }
 
+// LockProvider locks the specified amount of collaterals from a provider.
+// If necessary, it extends withdrawing collaterals and, if exist, their
+// linked unbondings as well.
+func (k Keeper) LockProvider(ctx sdk.Context, provider types.Provider, newLock types.LockedCollateral, lockPeriod time.Duration) {
+	originalCollateral := provider.Collateral
+	provider.Locked = provider.Locked.Add(newLock.Amount)
+	provider.Collateral = provider.Collateral.Sub(newLock.Amount)
+	provider.LockedCollaterals = append(provider.LockedCollaterals, newLock)
```

Comment:
Is it correct that we will still track individual lockings?

Reply:
I think we have to because otherwise we cannot calculate how much to take from each provider when we make a payout or unlock collaterals. The provider collateral to total collateral proportion fluctuates.